src/HordeAd/Core/Again.hs

```diff
@@ -830,6 +856,28 @@ example2 =
     (OS.fromList [1, 2, 3, 4] :: OS.Array [2, 2] Double)
     baz
 
+example2Record ::
+  ( HM.Numeric s,
+    KnownNat q,
+    KnownNat l,
+    KnownNat n,
+    KnownNat m,
+    KnownNat p
+  ) =>
+  ARecord
+    (OS.Array '[m, n] s)
+    (OS.Array '[n, p] s)
+    (OS.Array '[p, q] s) ->
+  OS.Array '[l, q] s ->
+  ( OS.Array '[l, q] s,
+    ARecord
+      (OS.Array '[m, n] s)
+      (OS.Array '[n, p] s)
+      (OS.Array '[p, q] s)
+  )
+example2Record (ARecord x y z) t =
+  runDualMonadAdapt (B.bipure ARecord ARecord <<*>> adaptArg x <<*>> adaptArg y <<*>> adaptArg z) t bazRecord
```

Comment:
Then you can get the gradient using this `Biapplicative` structure.  Further simplifications are possible but I wanted to know what you thought of this first.

Reply:
I guess I don't really understand because even if the original argument is an unboxed storable vector of scalars you will have to break those scalars out into (unboxed, unstorable) `D`s for the entirety of the differentiated program. So, I'm can't grasp why this is a useful thing to do.

Anyway, it doesn't really matter if I can't grasp it, because if your version can do it then I'm sure my version can be persuaded to do it too, and we can investigate that when appropriate.

Reply:
> That's why I don't store the collection of the initial dual numbers coming from parameters (the pairs of original value and the corresponding delta variable) in Ds, but in a pair of vectors

Hmm, isn't that what rank 1 is for?

Reply:
That's a good question. The only example for rank 0 I've seen is `prod` and it can be implemented on rank 1 as well. The `prod` benchmark could easily be extended with the rank 1 version but our benchmarks are broken anyway. I expect the rank 1 version to be much slower than rank 0 (with in-place updates), because vector indexing is slow (in-place updates are hard to implement there). However, perhaps we should just fix vector indexing, because it's so pervasive. There is a ticket about that, but it's outdated, because I can't get reliable benchmark data for it.

However, "pair of vectors" is beneficial even for rank 1 when you have many vectors and then "isn't that what rank 2 is for" may not apply when the vectors have different sizes. I don't remember how large the performance difference is in such cases (the vectors are both boxed, so it must be much less than for rank 0).

Reply:
I see. There is a lot of complexity here that I am not familiar with.